### PR TITLE
Dynamic Device addition and network discovery

### PR DESCRIPTION
**Pull request by Yashvardhan-A**
Network discovery introduced in the remote driver allows to find remote servers in LAN and WAN networks. mDNS through the Avahi library is used for LAN discovery and unicast-DNS-SD for discovery in specified domains. Using the OpenDHT library a DHT network can be established or joined where servers publish their information and clients can listen on the network to discover servers.

New enums for clGetDeviceInfo():
```
CL_DEVICE_REMOTE_SERVER_IP
CL_DEVICE_REMOTE_SERVER_PORT
```

New build options:
Remote Client-
```
ENABLE_REMOTE_DISCOVERY_AVAHI
ENABLE_REMOTE_DISCOVERY_DHT
```
Remote Server
```
ENABLE_REMOTE_ADVERTISEMENT_AVAHI
ENABLE_REMOTE_ADVERTISEMENT_DHT
```
New environment variables:
`POCL_DISCOVERY` of type bool. Introduced in device.h and used in device.c. The variable can be used to disable network discovery.

`POCL_REMOTE_SEARCH_DOMAINS` accepts a string, introduced in network_discovery.h and used in network_discovery.c. Used to specify the DNS domains in which queries should be made for unicast-DNS-SD. NULL means discovery will be performed only in LAN. By setting `POCL_REMOTE_SEARCH_DOMAINS="example.com" `discovery will be performed in LAN as well as in the specified domain.

`POCL_REMOTE_DHT_PORT` accepts an int, introduced in network_discovery.h, dht_advertise.h and used in network_discovery.c, dht_advertise.c. Used to specify a port to start the DHT node.

`POCL_REMOTE_DHT_BOOTSTRAP` accepts a string, introduced in network_discovery.h, dht_advertise.h and used in network_discovery.c, dht_advertise.c. Used to specify a bootstrap node which the current node will use to connect to a DHT network.

`POCL_REMOTE_DHT_KEY` accepts a string, introduced in network_discovery.h, dht_advertise.h and used in network_discovery.c, dht_advertise.c. After joining a DHT network, server and client nodes use a common key to publish or listen for server details.

In summary, this commit adds the following:

- Dynamic device addition
- Network discovery in PoCL-R
- Server-side advertisement in PoCL-D
- Reconnect and lost server handling in PoCL-R
- Enums in clGetDeviceInfo to get remote server IP and Port
- Documentation
- Test

### CODE REVIEW

File: include/CL/cl_ext_pocl.h [hunk 1]
```diff
@@ -147,12 +147,16 @@ typedef cl_int (CL_API_CALL *clSetKernelArgDevicePointerEXT_fn) (
 
 #define CL_DEVICE_REMOTE_TRAFFIC_STATS_POCL 0x4501
 
-/***********************************
-* cl_pocl_svm_rect +
-* cl_pocl_command_buffer_svm +
-* cl_pocl_command_buffer_host_buffer
-* extensions
-************************************/
+#define CL_DEVICE_REMOTE_SERVER_IP 0x4503
```
Comment (pjaaskel) [inline include/CL/cl_ext_pocl.h:150]:
```suggestion
#define CL_DEVICE_REMOTE_SERVER_IP_POCL 0x4503
```

Similar to the traffic stats constant.

File: lib/CL/devices/remote/network_discovery.c [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+  do                                                                          \
+    {                                                                         \
+      POCL_MEM_FREE (__OBJ__->id);                                            \
+      POCL_MEM_FREE (__OBJ__->domain);                                        \
+      POCL_MEM_FREE (__OBJ__->ip_port);                                       \
+    }                                                                         \
+  while (0)
+
+/**
+ * This structure contains remote server inforamtion about remote server found
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:64]:
```suggestion
 * This structure contains remote server information about remote servers found
```

File: lib/CL/devices/remote/network_discovery.c [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+ * through discovery. A hash table handle (`UT_hash_handle`) is included to
+ * allow instances of this struct to be managed in a hash table, preventing
+ * redundant entries when the same server is discovered multiple times.
+ *
+ * \brief Represents information about a remote server discovered on the
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:69]: Move this above the extended comment and drop \brief.

File: lib/CL/devices/remote/network_discovery.c [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+ * to a know server with same session.
+ */
+cl_int (*reconnect) (const char *);
+
+unsigned dev_type_idx;
+
+static server_info *server_table = NULL;
+static pocl_lock_t server_info_lock = POCL_LOCK_INITIALIZER;
+
+void
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:96]: static?

File: lib/CL/devices/remote/network_discovery.c [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+                 server_info *p_info)
+{
+  if (!device_count)
+    {
+      POCL_MSG_ERR ("Resolver called with zero devices.\n");
+      return;
+    }
+
+  server_info *new_server = p_info;
+  if (!new_server)
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:128]: Same here.
```suggestion
  if (new_server == NULL)
```
Strictly, I think NULL is allowed to be non-zero, although this is not likely the case in any existing systems.

File: lib/CL/devices/remote/network_discovery.c [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+   IN THE SOFTWARE.
+*/
+
+#include <stdatomic.h>
+#include <stdlib.h>
+#include <string.h>
+
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:29]: #include <stdbool.h>  (if you use bool in C)

File: lib/CL/devices/remote/network_discovery.c [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+                         dht_runner_get_bound_port (node, AF_INET));
+
+  return errcode;
+}
+#endif
+
+/*****************************************************************************/
+
+/**
+ * Function called by the remote driver to initialize network disocvery
```
Comment (pjaaskel) [inline lib/CL/devices/remote/network_discovery.c:630]:
```suggestion
 * Function called by the remote driver to initialize network discovery
```

File: lib/CL/pocl_cl.h [hunk 8]
```diff
@@ -510,6 +510,17 @@ struct pocl_device_ops {
    * NULL */
   cl_int (*reinit) (unsigned j, cl_device_id device, const char *parameters);
 
+  /**
+   * Initialize discovery mechanism in the driver to dynamically find new
+   * devices
+   * @param disco_dev_init_callback callback in runtime to init found devices
```
Comment (pjaaskel) [inline lib/CL/pocl_cl.h:516]:
```suggestion
   * \param disco_dev_init_callback callback in runtime to init found devices
```
Let's converge/standardize to \param.

File: pocld/avahi_advertise.c [hunk 9]
```diff
@@ -0,0 +1,267 @@
+/* avahi_advertise.c - part of pocl-daemon that performs mDNS on local network
```
Comment (pjaaskel) [inline pocld/avahi_advertise.c:1]: This is part of the server-side, so could have been C++ for more productive programming, but perhaps too late to convert.

File: pocld/dht_advertise.cc [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+/// Called by the daemon to start advertising using DHT using a seprate thread.
+///
+/// \param info contains the serialized data that is to be published on the DHT
+/// network. Data fields shared through info should be known to the client.
+
+void init_dht_advertisement(std::vector<uint8_t> info) {
+
+  dht::DhtRunner node;
```
Comment (pjaaskel) [inline pocld/dht_advertise.cc:63]: Check the naming caps, this is a C++ file.

**Review by pjaaskel — APPROVED** [review]

File: lib/CL/devices/remote/communication.c [hunk 11]
```diff
@@ -816,17 +890,36 @@ pocl_remote_reader_pthread (void *aa)
...
+              POCL_LOCK (inflight->mutex);
+              /* Each command in the inflight queue of the failed server has to
+               * be handled and marked as failed to prevent deadlock. */
+              DL_FOREACH (inflight->queue, cmd)
+                {
+                  DL_DELETE (inflight->queue, cmd);
+                  finish_running_cmd (cmd, NETCMD_FAILED);
+                }
+              POCL_UNLOCK (inflight->mutex);
+              printf("\nI was here\n");
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/communication.c:917]: Ah yes forgot to remove it.

File: lib/CL/devices/remote/network_discovery.c [hunk 12]
```diff
@@ -0,0 +1,656 @@
...
+                       strlen (new_server->ip_port), new_server);
+    }
+}
+
+/**
+ * The function takes the server information from the discovery resolvers and
+ * decides to add, reconnect or ignore the passed server and its devices.
+ */
+static void
+resolver (const char *id,
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:170]: Okay yes, renamed it to "handle_server_discovery"

File: lib/CL/devices/remote/network_discovery.c [hunk 13]
```diff
@@ -0,0 +1,656 @@
...
+{
+  POCL_LOCK (server_info_lock);
+
+  server_info *head = server_table, *p_info = NULL;
+
+  /* address && id -> reconnect normally
+   * !address && id -> ignore new address
+   * address && !id -> session changed, add as new service but need to
+   * handle repeatr address in find or create new server function !address &&
+   * !id -> add as a new service */
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:184]: yes, i have corrected for now. Once I finish all the edits, I'll check the comment again in case autoformatting messes it again.

**Comment by pjaaskel** [comment]
@franz can you add the missing libs to the LLVM 18 bot for testing this (see the failures)?

File: pocld/dht_advertise.cc (hunk 10 above)
Comment (Yashvardhan-A) [reply pocld/dht_advertise.cc:63]: It is a c++ file now.

**Comment by pjaaskel** [comment]
Needs rebasing and @franz to install the prerequisites.

File: lib/CL/devices/remote/network_discovery.c [hunk 14]
```diff
@@ -0,0 +1,656 @@
...
+
+  dht_runner *node = dht_runner_new ();
+  dht_runner_config dht_config;
+  dht_runner_config_default (&dht_config);
+
+  /* Port to start the DHT taken from environment or default used.*/
+  const int port = pocl_get_int_option (POCL_REMOTE_DHT_PORT, 4222);
+  /* DHT bootstrap node taken from environment or default used.*/
+  const char *bootstrap
+    = pocl_get_string_option (POCL_REMOTE_DHT_BOOTSTRAP, "bootstrap.jami.net");
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c]: bootstrap.jami.net is the open public node used to connect to a public DHT network ([Jami distributed network](https://docs.jami.net/en_US/user/jami-distributed-network.html#jami-distributed-network)). I am not sure about the exact terms of use, but it is kept running by many stable public nodes and seems to be open for anyone to use.

File: lib/CL/devices/remote/network_discovery.c [hunk 15]
```diff
@@ -0,0 +1,687 @@
...
+
+  /* Port to start the DHT taken from environment or default used.*/
+  const int port = pocl_get_int_option (POCL_REMOTE_DHT_PORT_ENV, 4222);
+  /* DHT bootstrap node taken from environment or default used.*/
+  const char *bootstrap
+    = pocl_get_string_option (POCL_REMOTE_DHT_BOOTSTRAP_ENV, NULL);
+  if (bootstrap == NULL)
+    {
+      POCL_MSG_ERR ("DHT bootstrap node environment variable not specified, "
+                    "server advertisement failed! \n");
```
Comment (jansol) [inline lib/CL/devices/remote/network_discovery.c:629]: This is client code, no?

Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:629]: Ah yes, I copied the comment from the server and forgot to modify.

File: pocld/daemon.cc [hunk 16]
```diff
@@ -66,6 +70,10 @@
     }                                                                          \
   } while (0)
 
+#ifdef ENABLE_REMOTE_ADVERTISEMENT_DHT
+std::thread DHTThread;
```
Comment (jansol) [inline pocld/daemon.cc:74]: should be a member of the PoclDaemon class, not a global.

File: pocld/daemon.cc [hunk 17]
```diff
@@ -302,6 +556,14 @@ int PoclDaemon::launch(std::string ListenAddress, struct ServerPorts &Ports,
     POCL_MSG_ERR("Error resolving address: %s\n", gai_strerror(error));
     return -1;
   }
+
+#ifdef ENABLE_REMOTE_ADVERTISEMENT_AVAHI
+  avahiAdvertiseP = StartAvahiAdvert(ResolvedAddress, Ports);
```
Comment (jansol) [inline pocld/daemon.cc:561]: Not necessary to do in this PR now, but you *could* use std::unique_ptr since nothing else ever touches this pointer. Then you can assign it here with `variablename.reset(StartAvahiAdvert(ResolvedAddress, Ports));` and don't need to do anything manually in the PoclDaemon destructor.

File: lib/CL/devices/remote/network_discovery.c [hunk 18]
```diff
@@ -0,0 +1,687 @@
...
+
+  clear_avahi ();
+  return errcode;
+}
+#endif
+
+/*****************************************************************************/
+#if defined(ENABLE_REMOTE_DISCOVERY_DHT)
+
+const char *common_key = NULL;
```
Comment (jansol) [inline lib/CL/devices/remote/network_discovery.c:522]: Should this really be a global variable?

**Review by jansol — APPROVED** [review]
A few variables look like they are more widely visible than they should be, with those fixed LGTM.

File: lib/CL/devices/remote/network_discovery.c [hunk 19]
```diff
@@ -0,0 +1,687 @@
...
+#include <avahi-client/lookup.h>
+#include <avahi-common/domain.h>
+#include <avahi-common/error.h>
+#include <avahi-common/malloc.h>
+#include <avahi-common/thread-watch.h>
+
+AvahiThreadedPoll *avahi_threaded_poll = NULL;
+AvahiClient *avahi_client = NULL;
+cl_int domain_count = 0;
+AvahiServiceBrowser **service_browser = NULL;
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:48]: Yeah they should be static.

File: lib/CL/devices/remote/network_discovery.c [hunk 20]
```diff
@@ -0,0 +1,687 @@
...
+ * add and initialize devices in the discovered remote server.
+ */
+cl_int (*add_discovered_device) (const char *, unsigned);
+
+/* Function pointer for the callback defined in pocl-r driver to reconnect
+ * to a know server with same session.
+ */
+cl_int (*reconnect) (const char *);
+
+unsigned dev_type_idx;
```
[the remark this answers is not in the document]
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:91]: These should be static as well.

File: lib/CL/devices/remote/network_discovery.c (hunk 18 above)
Comment (Yashvardhan-A) [reply lib/CL/devices/remote/network_discovery.c:522]: Only for this file, therefore also static.

File: pocld/daemon.cc (hunk 17 above)
Comment (Yashvardhan-A) [reply pocld/daemon.cc:561]: Ah yes, that's cool!

File: pocld/daemon.cc (hunk 16 above)
Comment (Yashvardhan-A) [reply pocld/daemon.cc:74]: I have resolved this using a friend function for now. Not sure if that's the 'correct' way to do it.

File: lib/CL/devices/devices.c [hunk 21]
```diff
@@ -511,6 +515,103 @@ pocl_reinit_devices ()
   return retval;
 }
 
+/**
+ * Callback function that dynamically adds and initializes devices. It is
```
Comment (pjaaskel) [inline lib/CL/devices/devices.c:519]: For the next PR: Split the brief description and the extended description with an empty line.

**Comment by pjaaskel** [comment]
I think this is now only waiting for a rebase and @franz to add required libs to the CI bot.

**Comment by franz** [comment]
avahi/dht libraries have been added to the CI

**Comment by pjaaskel** [comment]
Thanks!
